pkg/util/ranger/detacher.go

```diff
@@ -273,8 +273,37 @@ func extractBestCNFItemRanges(sctx *rangerctx.RangerContext, conds []expression.
 			continue
 		}
 		curRes := getCNFItemRangeResult(sctx, res, i)
-		if bestRes == nil || compareCNFItemRangeResult(curRes, bestRes) {
+		if bestRes == nil {
 			bestRes = curRes
+		} else {
+			if curRes != nil && bestRes.rangeResult != nil {
+				if fixcontrol.GetBoolWithDefault(sctx.OptimizerFixControl, fixcontrol.Fix16798, false) {
+					bestResIsSubset := bestRes.rangeResult.Ranges.Subset(sctx.TypeCtx, curRes.rangeResult.Ranges)
+					// if bestRes is a subset then do nothing
+					if !bestResIsSubset {
+						curResIsSubset := curRes.rangeResult.Ranges.Subset(sctx.TypeCtx, bestRes.rangeResult.Ranges)
+						// if curRes is subset (more selective) then make it bestRes.
+						if curResIsSubset {
+							bestRes = curRes
+						} else {
+							// Try intersecting result of different conjuncts.
+							intersection := curRes.rangeResult.Ranges.IntersectRanges(sctx.TypeCtx, bestRes.rangeResult.Ranges)
+							// Skip intersection if an error occurred during the intersection computation.
+							if intersection == nil {
+								if compareCNFItemRangeResult(curRes, bestRes) {
+									bestRes = curRes
+								}
+							} else {
+								bestRes.rangeResult.Ranges = intersection
+								bestRes.rangeResult.AccessConds =
+									AppendConditionsIfNotExist(bestRes.rangeResult.AccessConds, []expression.Expression{cond})
+							}
+						}
+					}
+				} else if compareCNFItemRangeResult(curRes, bestRes) {
+					bestRes = curRes
+				}
+			}
```

Comment:
I think we can make the entire logic of getting the best from `curRes` and `bestRes` a separate function.
That should make the new function have a clear input and output, and also reduce the deep indentations, also looks like the repeated `compareCNFItemRangeResult` can be removed.

Reply:
Makes sense. Done